### ISSUE TITLE
Delegate Order

### ISSUE THREAD

**Issue by jonmifsud** [symphonycms/symphony-2#2354]
Some extensions such as [Author Roles](http://symphonyextensions.com/extensions/author_roles/) potentially depend on the order in which Delegates are run. Author roles in-fact asks for the users to edit/touch the core to ensure that it's the last thing to run.

I'm not very keen on touching up the core for every project, If it makes sense we should include an 'order' field within the `tbl_extensions_delegates` defaulting to 0 and add another intermediary step in [the sorting](https://github.com/symphonycms/symphony-2/blob/master/symphony/lib/toolkit/class.extensionmanager.php#L66). That way if an extension would like to be after the 'standard' extensions it would need to define a value larger than 0. If it doesn't care about which order it is run, no code would need to change as value would default to 0, as they will still be sorted alphabetically after that.

**Comment by nitriques** [comment]
Jonathan, this already has been suggested. I am not 100% against it, but it would need core UI in order to change the order. Having two way to order is also not a good idea: If a order column exists, it should always be used (and kept synced).

@brendo @michael-e @nilshoerrmann Do you remember the conclusion we made on that topic ?

**Comment by jonmifsud** [comment]
To be fair I've been poking around the core, seems like since the "Manager" permission level was added, there's a way to add/remove items from navigation in a much easier way. Seems like there's [a function](https://github.com/symphonycms/symphony-2/blob/master/symphony/lib/toolkit/class.administrationpage.php#L520) which check if the user has permissions before showing the respective links.

I think if there's a delegate there, it would make more sense rather than allowing the navigation to build and then hiding items. And would resolve the mentioned issue, at least for this use case.

**Comment by nitriques** [comment]
Adding a delegate in a utility method is not a good idea since this can be called by anybody.

But you could look at another spot to add a delegate (or use this one https://github.com/symphonycms/symphony-2/blob/master/symphony/lib/toolkit/class.administrationpage.php#L784 ?)

**Comment by jonmifsud** [comment]
@nitriques that's where the Author Roles does it's magic right now. I didn't check how many other extensions I'm using plug into that right now. But the fact that it asks to modify the core so it can give appropriate results is somewhat strange.

> this can be called by anybody

maybe that is what would make it more appropriate in terms of permissions. I didn't check from where it's called but I guess that it could be possible to block access to extensions and section pages because of user permissions, even if someone has a direct link. (though `canAccessPage` might be more appropriate than `doesAuthorHaveAccess`)

Either way you know the system better than I do.

**Comment by nitriques** [comment]
> But the fact that it asks to modify the core so it can give appropriate results is somewhat strange.

I agree.

`doesAuthorHaveAccess` really is a utility method. Delegates should only be placed in Symphony's life cycle methods. 

Would creating a `NavigationPostRender` deleage here worked ? https://github.com/symphonycms/symphony-2/blob/master/symphony/lib/toolkit/class.administrationpage.php#L838

**Comment by brendo** [comment]
> @brendo @michael-e @nilshoerrmann Do you remember the conclusion we made on that topic ?

It was closed as a Symphony 2 issue which would be resolved by Symphony Next. I considered it an edge case and the effort could be used elsewhere.

I like @jonmifsud's suggestion, and I think it could be done without requiring any new UI.

- The `tbl_extensions_delegates` would be extended to have an additional column, `order`, with the default value of 0. This would be an unsigned column to allow for negative values.
- When extensions provide their delegates (`getSubscribedDelegates`), they can optionally include a `order` key. If this is not provided, it would be 0 (normal).
- When a delegate is resolved, extensions will be ordered by the `order` column first, and then alphabetically. This would mean that the existing delegate execution order is preserved, but extensions that also provide an order would get the updated benefits.

With the above outlined, this would not be a breaking change.

**Comment by brendo** [comment]
Playing around with this, it actually is stupidly simple to implement as MySQL takes care of the sorting for us and for the most part, it's 'opt in' only. I've created a new branch for this feature if you'd like to test @jonmifsud. Simply add the `order` column to the table, and then an `order` key to the delegate subscription in the affected extensions.

`0` is considered default. Negative orders will be run first, larger orders last.

Let me know how it works for you :)

**Comment by nitriques** [comment]
> This would be an unsigned column to allow for negative values.

Did you mean signed ?

So, if I understand correctly, the only ways to change the order value is

1. on extension install
2. manually via the DB

?

**Comment by brendo** [comment]
Yep, signed, sorry!

Yes, it's done by the extension, not the developer. That's not to say a UI
couldn't be put on it, but I think it's overly complicating an already
uncommon use case

**Comment by nitriques** [comment]
Ok I agree with you.

But how, as an extension developer, will I be able to set the order at install time ? We need an API for that because right now [install](https://github.com/symphonycms/symphony-2/blob/master/symphony/lib/toolkit/class.extensionmanager.php#L362) is called before the [we insert the row into the DB](https://github.com/symphonycms/symphony-2/blob/master/symphony/lib/toolkit/class.extensionmanager.php#L387).

**Comment by jonmifsud** [comment]
Right I guess it would be something you have to set within the delegates
definition.

**Comment by nitriques** [comment]
> within the delegates definition.

I was thinking about the same thing. A 'post-install' delegate would be nice.

**Comment by jonmifsud** [comment]
I was thinking more on the lines, as you set the delegate which you want to
fire, and the callback funciton, there could be an optional 'priority'
field. rather than a post-install delegate.

**Comment by nitriques** [comment]
>  rather than a post-install delegate.

Hum that's interesting too. But @brendo's idea was to put the order in the DB on a per extension basis, not on a per delegate basis.

Putting the priority in the delegate "declaration" would require sorting the delegates in PHP, which may not be super awesome. + It's not editable.

**Comment by brendo** [comment]
> Hum that's interesting too. But @brendo's idea was to put the order in the DB on a per extension basis, not on a per delegate basis.

Not quite, extension developers will add the order to the delegate subscription that lives in `getSubscribedDelegates`.

>> When extensions provide their delegates (getSubscribedDelegates), they can optionally include a order key. If this is not provided, it would be 0 (normal).

> Putting the priority in the delegate "declaration" would require sorting the delegates in PHP, which may not be super awesome. + It's not editable.

It can be editable... in a sense, because the value is copied into the extension delegates table. The theoretical UI could change the database values. The one potential problem with this is that extension reinstalls would revert the order back to whatever the extension defined.

I think the key function that you both missed is [`registerDelegates`](https://github.com/symphonycms/symphony-2/blob/master/symphony/lib/toolkit/class.extensionmanager.php#L499-L532), which is called during the extension installation. Take a look at [this line](https://github.com/symphonycms/symphony-2/blob/master/symphony/lib/toolkit/class.extensionmanager.php#L512), it invokes the `getSubscribedDelegates` method on the extension object and adds that data to the extension delegates table.

**Comment by jensscherbl** [comment]
Mhhh. I don't wanna spoil your party, but wouldn't ordering with numbers only solve part of the problem? What if an extensions needs to run its code for a specific delegate before a certain extension, but after another one? I agree that numbers are better than nothing, but it would still require all developers of such edge case extensions to talk to each other and get the numbering right.

**Comment by nitriques** [comment]
Ok so we are ordering delegates, not extensions.. Makes sense.

**Comment by brendo** [comment]
The aformentioned branch has been merged into `integration`.

**Comment by nitriques** [comment]
Great!

**Comment by jensscherbl** [comment]
> To be fair, I can't think of a better solution either, so @brendo's proposal kinda makes sense.

What if extensions could decide to run after specific other extensions from inside a delegate callback?

**Comment by nitriques** [comment]
> What if extensions could decide to run after specific other extensions from inside a delegate callback?

It's dangerous to create dependency circles. A needs to wait for B, B waits for A. How would you break that circle ?

**Comment by jensscherbl** [comment]
> It's dangerous to create dependency circles. A needs to wait for B, B waits for A. How would you break that circle ?

Why would `B `wait for `A` if `A` has to run after `B`?

It's just a more flexible solution for the same problem. It's an optional mechanism for extension developers to run their callbacks after another, specific extension.

Problem with the numbers is, if `A` has to run after `B` and sets its number to 1 (because `B` uses the default of 0), the entire chain breaks if `C` and `D` come along and `B` is updated to run after `D`.

**Comment by nitriques** [comment]
> Why would B wait for A if A has to run after B?

Because somebody made a mistake ;)

> It would require a massive amount of communication and coordination between extension developers to keep everything in the right order as time goes by.

Yes you're right.

> If A could just say "I need to run after b" instead, it wouldn't have to care about B's dependencies and if its number is still high enough to ensure the right order.

Yeah but the core needs to handle possible error cases... if not, you'll get an infinite loop.

**Comment by brendo** [comment]
> What if extensions could decide to run after specific other extensions from inside a delegate callback?

So, Extension Elephants has it's own delegate, `GetPeanuts` that is fired after the `EntryPostCreate` delegate? Extension Circus would register `GetPeanuts` in their own `getSubscribedDelegates` method. This functionality is available now. I'm not sure how ordering plays a part in this?

Or are you talking more general, Extension Car wants to run when `EntryPostCreate` is triggered, but only after Extension Mechanic has done their bit on `EntryPostCreate`? This would be a case for ordering. It can be solved by Extension Car uses a lower order than Extension Mechanic.

By default all extensions order is set to 0. This gives extensions the freedom to play with the order by going 'higher' or 'lower'. So if Extension Mechanic didn't specify an order, they would default to `0`. Extension Car can then set their order at `-16` to ensure they run after Extension Mechanic. While they can also just set `-1`, it doesn't give other developers much freedom to run inbetween. So for example if Extension MechanicManager wanted to run after Mechanic but before Car, they could set their order to be `-8`.

Yes. I realise that this still has a flaw in that Extension Car had to have the foresight to use `-16` to give this buffer, but I'm really sure how else you'd do this. Given the use case for wanting this delegate ordering is rare at the moment, I don't think we have to over engineer this functionality at this point.

The ultimate solution is what has been previously suggested, create a UI where the developer can log in and move extension delegates into their own order. There's something about this I don't like, because now the developer must have intimate knowledge of both extensions (or more) to execute them in a particular order. This may lead to further bugs. What if Extension A expected a parameter to be an object and in the usual execution order, it is, but now a developer has moved Extension C to execute first, and that extension actually is a bit naughty and turns the parameter into a string. Suddenly Extension A doesn't work anymore. 

Maybe I'm just being paranoid.

**Comment by nitriques** [comment]
> Maybe I'm just being paranoid.

I do not think you are. People make PhDs on that subject every year. Managing dependancies is a NP-hard problem.

I would put emphasis on:

>  I don't think we have to over engineer this functionality at this point.

Totally. The simpler the better.

**Comment by jensscherbl** [comment]
**tl;dr**

> I don't think we have to over engineer this functionality at this point

+1

> So, Extension Elephants has it's own delegate, GetPeanuts that is fired after the EntryPostCreate delegate? Extension Circus would register GetPeanuts in their own getSubscribedDelegates method. This functionality is available now.

Already used this approach in some projects. But this would require extension developers to introduce their own delegates. Therefore have the foresight that their extension might cause problems in certain scenarios and that other extensions might have to run after your own extension.

The concept I suggested would reverse this responsibility.

> Or are you talking more general, Extension Car wants to run when EntryPostCreate is triggered, but only after Extension Mechanic has done their bit on EntryPostCreate?

Exactly.

> Given the use case for wanting this delegate ordering is rare at the moment, I don't think we have to over engineer this functionality at this point.

Fine with me. Just wanted to throw my two cents in to make sure it doesn't backfire right away.

> The ultimate solution is what has been previously suggested, create a UI where the developer can log in and move extension delegates into their own order. 

Not a solution at all because of the problems you've mentioned. I'd be against this.

**Comment by brendo** [comment]
> Not a solution at all because of the problems you've mentioned. I'd be against this.

Did you have an idea of how else this could be achieved?

It seems to always boil down to two things. Either the extension developers talk, or the developer takes control. Looking at Symfony's and Laravel, they both place the onus on the developer. The developer is responsible for setting the priority of event registration.

**Comment by jensscherbl** [comment]
> Looking at Symfony's and Laravel, they both place the onus on the developer.

Those are php frameworks for php developers. Symphony developers aren't necessarily experienced php developers, nor should they have to understand the inner workings of each and every extension to make an informed decision about the right execution order.

> Did you have an idea of how else this could be achieved?

Maybe if such dependencies could be documented in an easy to understand and standard way for all extensions.

But again, could be an edge case and leaving enough buffer between numbers sounds like a pragmatic and reasonable enough workaround.

**Comment by michael-e** [comment]
I [commented on the commit](https://github.com/symphonycms/symphony-2/commit/b4dca4a7e5399b3e22b2cf74711a23804e04238f#commitcomment-10257829).
